types/types.go

```diff
@@ -353,6 +353,70 @@ type Alert struct {
 	Timeout   bool
 }
 
+type AlertSnapshot struct {
```

Comment:
I had a quick discussion with @gotjosh about trying to make this "immutable" by making the fields package-private and providing Getters for each of them. The problem is that Go does not really support immutable types, and in this example both the `Labels()` and `Annotations()` return a reference to the same memory, so those are actually mutable.

I think it would be more idiomatic to just have:

```go
type AlertSnapshot struct {
    Alert
    ref time.Time
}
```

and then override the required methods such as:

```go
func (a *AlertSnapshot) Resolved() bool {
    return a.ResolvedAt(a.ref)
}
```

We can document that this struct is to be treated as immutable if we need to pass references to it.

Reply:
Changed.